File: 06-09-2024_Kadane's_Algorithm.java
class Solution {

    // arr: input array
    // Function to find the sum of contiguous subarray with maximum sum.
    long maxSubarraySum(int[] arr) {

        // Your code here
        long maxSoFar = Integer.MIN_VALUE, maxEndingHere = 0;

        for (int i = 0; i < arr.length; i++) {
            maxEndingHere += arr[i];

            // Use Math.max to find the maximum 
            // of maxSoFar and maxEndingHere
            maxSoFar = Math.max(maxSoFar, maxEndingHere);

            // Use Math.max to reset maxEndingHere 
            // to 0 if it becomes negative
            maxEndingHere = Math.max(maxEndingHere, 0);
        }
        return maxSoFar;
        
    }
}

File: 06-09-2024_Kadane's_Algorithm.cpp
class Solution {
  public:
    // Function to find the sum of contiguous subarray with maximum sum.
    long long maxSubarraySum(vector<int> &arr) {
        // code here...
        long long max_so_far = -1e12, max_ending_here = 0;

        for (int i = 0; i < arr.size(); i++) {      
            max_ending_here = max_ending_here + arr[i];
          
            // If max_ending_here becomes more than 
            // the current result
            max_so_far = max(max_so_far, max_ending_here);
           
            // If max_ending_here becomes negative
            max_ending_here = max(max_ending_here, 0LL);
        }
        
        return max_so_far;
        
    }
};
